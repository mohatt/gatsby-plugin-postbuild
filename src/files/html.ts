import { Promise } from 'bluebird'
import path from 'path'
import _ from 'lodash'
import { DefaultTreeAdapterTypes as parse5, parse, serialize, defaultTreeAdapter } from 'parse5'
import File, { FileConstructorArgs } from './base'
import type { IExtensionOptions } from '../interfaces'

export type IFileHtmlOptions = IExtensionOptions<{
  commons: {
    [type: string]: string[]
  }
}>

const DEFAULTS = {
  commons: {
    style: ['gatsby-global-css'],
    script: ['gatsby-chunk-mapping'],
  },
}

/**
 * A callback function to be run on tree nodes
 * @see FileHtml.walk
 */
export type IFileHtmlNodeWalker = (
  node: parse5.Node,
  prev?: parse5.Node,
  next?: parse5.Node,
) => Promise<void> | void

/**
 * Interface for html files
 */
export class FileHtml extends File {
  /**
   * Path to the html page
   */
  pagePath: string

  /**
   * Compiled ast tree
   */
  tree: parse5.Document

  /**
   * Parse5 tree adaptor
   */
  adaptor = defaultTreeAdapter

  /**
   * Html extension options
   */
  declare options: IFileHtmlOptions

  /**
   * Sets the page path and creates an empty parse5 document
   * @internal
   */
  constructor(rel: string, options: IFileHtmlOptions, args: FileConstructorArgs) {
    super(rel, options, args)

    // @todo options should be loaded once
    this.options = _.defaultsDeep(options, DEFAULTS)

    // Set the path to the html page
    const parts = rel.slice(0, -5).split(path.sep)
    const l = parts.length - 1
    if (parts[l] === 'index') parts.pop()
    else if (parts[l] === '404') parts[l] = '404.html'
    parts.unshift(args.gatsby.pathPrefix)
    this.pagePath = parts.join('/') || '/'

    // Create an empty document
    this.tree = this.adaptor.createDocument()
  }

  /**
   * Reads the file then compiles its html to ast
   * @internal
   */
  read(): Promise<void> {
    const payload = this.emitPayload<FileHtml>()
    return this.file
      .read()
      .then((html) =>
        this.emit(
          'html',
          'parse',
          {
            ...payload,
            html,
          },
          'html',
        ),
      )
      .then((html) => {
        this.tree = parse(html)
        return this.emit('html', 'tree', payload)
      }) as Promise<void>
  }

  /**
   * Invokes html.node events on all nodes
   * @internal
   */
  process(): Promise<void> {
    return this.walk((node, previousNode, nextNode) => {
      return this.emit('html', 'node', {
        ...this.emitPayload<FileHtml>(),
        node,
        previousNode,
        nextNode,
      }) as unknown as Promise<void>
    })
  }

  /**
   * Serializes the tree back to html and writes the file
   * @internal
   */
  write(): Promise<void> {
    const payload = this.emitPayload<FileHtml>()
    return this.emit('html', 'serialize', payload)
      .then(() => {
        const html = serialize(this.tree)
        return this.emit(
          'html',
          'write',
          {
            ...payload,
            html,
          },
          'html',
        )
      })
      .then((html) => this.file.update(html))
  }

  /**
   * Recursively walks through a parse5 tree invoking a callback
   *  on every node on the tree.
   */
  walk(
    cb: IFileHtmlNodeWalker,
    root: parse5.Node = this.tree,
    processGatsbyNode = false,
  ): Promise<void> {
    let gatsby = processGatsbyNode
    const walker = async (
      curr: parse5.Node,
      prev?: parse5.Node,
      next?: parse5.Node,
    ): Promise<void> => {
      // Disallow processing `___gatsby` node and its descendants by default
      if (
        !gatsby &&
        'attrs' in curr &&
        curr.attrs.find((a) => a.name === 'id')?.value === '___gatsby'
      ) {
        gatsby = true
        return
      }
      await cb(curr, prev, next)
      if ('childNodes' in curr) {
        await Promise.each(curr.childNodes, (childNode, i) => {
          return walker(childNode, curr.childNodes[i - 1], curr.childNodes[i + 1])
        })
      }
    }
    return walker(root)
  }

  /**
   * Resolves a href attribute of a given node into local file path
   *
   * @todo test this method on win32 platform
   * @param node - Parse5 element node
   * @param attrib - Target attribute. Defaults to `href`
   * @param relative - Return path relative to `/public`. Defaults to `true`
   * @param strict - Return false if resolved path is outside `/public`. Defaults to `true`
   * @return Resolved file path or false on failure to meet conditions
   */
  resolveHref(
    node: parse5.Element,
    attrib = 'href',
    relative = true,
    strict = true,
  ): string | boolean {
    const { filesystem, gatsby } = this.emitPayload()
    const attr = node.attrs.find((a) => a.name === attrib)
    if (attr === undefined) return false
    let href = attr.value.trim()
    if (href === '' || /^\w+:\/\//.test(href)) return false
    const prefix = gatsby.pathPrefix
    if (prefix !== '') {
      if (href.indexOf(prefix) === 0) {
        href = href.replace(prefix, '')
        // absolute path oustide gatsby root
      } else if (path.isAbsolute(href)) {
        return false
      }
    }

    const absPath = path.isAbsolute(href)
      ? path.join(filesystem.root, href)
      : path.resolve(path.dirname(this.path), href)
    const relPath = path.relative(filesystem.root, absPath)
    if (strict && relPath.indexOf('..') === 0) return false
    return relative ? relPath : absPath
  }
}

export default FileHtml
