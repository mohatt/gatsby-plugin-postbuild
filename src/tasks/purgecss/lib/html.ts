import { Promise } from 'bluebird'
import crypto from 'crypto'
import * as parse5 from 'parse5'
import { cloneDeep } from 'lodash'
import type { Filesystem, FileHtml } from '@postbuild'
import type { Purger, IPurgeResult } from './purger'
import type AssetMapper from './mapper'
import type IOptions from '../options'

/**
 * Utility function to create an id for an inline
 * style element that is shared between multiple html
 * files
 */
function createStyleId (css: string): string {
  return crypto.createHash('sha1')
    .update(css, 'utf8')
    .digest('base64')
}

/**
 * Interface for style objects
 */
export abstract class HtmlStyle {
  /**
   * Style ID used for linking styles to html files
   */
  protected id?: string
  getID = (): string => this.id !== undefined ? this.id : ''
  hasID = (): boolean => this.id !== undefined
  setID = (id: string): void => {
    this.id = id
  }

  /**
   * Reads the original css string
   */
  abstract read (): Promise<string>|string

  /**
   * Writes the optimized css string
   */
  abstract update (result: IPurgeResult): Promise<void>|void
}

/**
 * Handles inline style nodes
 */
export class HtmlStyleInline extends HtmlStyle {
  private textNode: parse5.TextNode
  constructor (node: parse5.TextNode) {
    super()
    this.textNode = node
  }

  read (): string {
    return this.textNode.value
  }

  update (result: IPurgeResult): void {
    this.textNode.value = result.css
  }
}

/**
 * Handles external css files
 */
export class HtmlStyleFile extends HtmlStyle {
  private readonly fs: Filesystem
  private readonly options: IOptions
  private readonly path: string
  constructor (path: string, options: IOptions, fs: Filesystem) {
    super()
    this.fs = fs
    this.options = options
    this.path = path
  }

  read (): Promise<string> {
    return this.fs.read(this.path)
  }

  update (result: IPurgeResult): Promise<void> {
    return this.fs.update(this.path, result.css, {
      purgecss: String(result.rejected.length)
    }).then(() => {
      if (this.options.writeRejected) {
        return this.fs.create(this.path + '.rejected.txt', result.rejected.join(' '))
      }
    })
  }
}

/**
 * Handles html parsing/serialization and asset searching
 */
export class HtmlOptimizer {
  /**
   * File object received from postbuild
   */
  readonly file: FileHtml

  /**
   * Html string without inline styles
   */
  html: string = ''

  /**
   * List of scripts found in html
   */
  readonly scripts: string[] = []

  /**
   * List of styles found in html with their metadata
   */
  private readonly styles: HtmlStyle[] = []

  private readonly options: IOptions
  private readonly fs: Filesystem
  private readonly mapper: AssetMapper
  private readonly purger: Purger
  private _inHead = true

  /**
   * Creates a new instance for the given html file
   */
  constructor (file: FileHtml, options: IOptions, fs: Filesystem, mapper: AssetMapper, purger: Purger) {
    this.options = options
    this.file = file
    this.fs = fs
    this.mapper = mapper
    this.purger = purger
  }

  /**
   * Clones the ast tree and creates a style-ess
   * copy of it
   */
  async load (): Promise<void> {
    // clone the tree and remove all <style> nodes
    const purgedTree = cloneDeep(this.file.tree)
    await this.file.walk(node => {
      if (node.nodeName === 'style') {
        this.file.adaptor.detachNode(node)
      }
    }, purgedTree)
    // then serialize it back to HTML
    this.html = parse5.serialize(purgedTree)
  }

  /**
   * Searches for css/js assets in ast nodes
   */
  processNode (node: parse5.Node): void {
    if (node.nodeName === 'head') {
      this._inHead = true
      return
    }
    if (node.nodeName === 'body') {
      this._inHead = false
      return
    }
    if (node.nodeName === 'style') {
      return this.addStyle(node)
    }
    if (node.nodeName === 'link') {
      const rel = node.attrs.find(a => a.name === 'rel')
      if (rel?.value.toLowerCase().trim() !== 'stylesheet') return
      return this.addStyle(node)
    }
    if (node.nodeName === 'script') {
      this.addScript(node)
    }
  }

  /**
   * Creates a style object for the given <style> or <link> node
   * then appends it to {self.styles}
   */
  addStyle (node: parse5.Element): void {
    if (node.nodeName === 'link') {
      const stylePath = this.file.resolveHref(node)
      if (typeof stylePath !== 'string' ||
        this.fs.extension(stylePath, 'css') === false ||
        this.mapper.shouldIgnoreFile(stylePath)) {
        return
      }
      const style = new HtmlStyleFile(stylePath, this.options, this.fs)
      style.setID(stylePath)
      this.mapper.linkStyleToHtml(stylePath, this)
      this.styles.push(style)
      return
    }

    if (node.childNodes.length === 0) return
    const style = new HtmlStyleInline(node.childNodes[0] as parse5.TextNode)
    if (this._inHead) {
      const idAttr = node.attrs.find(a => a.name === 'id')
      const id = idAttr !== undefined && idAttr.value !== ''
        ? idAttr.value
        : createStyleId(style.read())
      style.setID(id)
      this.mapper.linkStyleToHtml(id, this)
    }
    this.styles.push(style)
  }

  /**
   * Creates a script filename from the given <script> node
   * checks if the script should be ignored, then appends
   * it to {self.scripts}
   */
  addScript (node: parse5.Element): void {
    const scriptPath = this.file.resolveHref(node, 'src')
    if (typeof scriptPath !== 'string' ||
      this.fs.extension(scriptPath, 'js') === false ||
      this.mapper.shouldIgnoreFile(scriptPath)) {
      return
    }
    this.scripts.push(scriptPath)
  }

  /**
   * Purges all style nodes
   */
  purgeStyles (): Promise<void> {
    return Promise.map(this.styles, style => {
      return this.purger.purge(style, this)
    }).then(result => {
      const rejected = result.flat()
      this.file.reportMeta.purgecss = String(rejected.length)
      if (this.options.writeRejected) {
        return this.fs.create(this.file.relative + '.rejected.txt', rejected.join(' '))
      }
    })
  }
}

export default HtmlOptimizer
