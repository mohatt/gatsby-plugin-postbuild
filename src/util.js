import path from 'path'
import crypto from 'crypto'
import _ from 'lodash'
import Debug from 'debug'

/**
 * Global options object
 *
 * @type {Object}
 */
export let options = {
  _plugin: 'gatsby-plugin-postbuild'
}

/**
 * Initializes plguin options and other utilities
 *
 * @param {Object} $0
 * @param   {Object} $0.defaultOptions - Options defaults
 * @param   {Object} $0.pluginOptions - Plugin options
 * @param   {Object} $0.gatsby - Gatsby's onPreBootstrap event object
 */
export function bootstrap ({ defaultOptions, pluginOptions, gatsby }) {
  // Merge user-defined options with defaults
  options = _.mergeWith(
    options,
    defaultOptions,
    pluginOptions,
    // Ensure arrays aren't merged by index
    (to, from) => _.isArray(from) ? from : undefined
  )
  // set private options
  options._root = gatsby.store.getState().program.directory
  options._public = path.join(options._root, 'public')
  options._pathPrefix = gatsby.pathPrefix

  // Initialize reporter
  const reporter = gatsby.reporter
  const errorMap = {
    10000: {
      text: context => `${options._plugin} threw an error while running:\n ${context.message}`,
      level: 'ERROR',
      type: 'PLUGIN'
    }
  }
  if (reporter.setErrorMap) {
    reporter.setErrorMap(errorMap)
  }
  debug('Plugin initialized', options)
}

/**
 * Creates a new structured error to pass to gatsby's reporter
 *
 * @constructor
 * @param {string|Object} message
 * @param {Error|null} error
 * @param {number} code
 * @return {{ id: string, context: {Object}, error: error}}
 */
export function ReporterError (message, error = null, code = 10000) {
  return {
    id: `${options._plugin}_${code}`,
    context: typeof message === 'object'
      ? message
      : { message },
    error: error && typeof error === 'object'
      ? error
      : undefined
  }
}

/**
 * Prints a debug message
 *
 * @param {string} message
 * @param {any} data
 */
export function debug (message, data = null) {
  createDebug()(...arguments)
}

/**
 * Creates a debug instance for the given namespace
 *
 * @param {string} namespace
 * @return {Function}
 */
export function createDebug (namespace = '') {
  let ns = options._plugin
  if (namespace) {
    ns += ':' + namespace
  }

  createDebug.cache ??= new Map()
  if (createDebug.cache.has(ns)) {
    return createDebug.cache.get(ns)
  }

  const realDebug = Debug(ns)
  const debug = function (message) {
    arguments[0] = arguments.length > 1
      ? `${message} %O \n\n`
      : `${message} \n\n`
    realDebug(...arguments)
  }

  createDebug.cache.set(ns, debug)
  return debug
}

/**
 * Creates a sha1 hash from a string.
 *
 * @param {string} data
 * @param {boolean} base64
 */
export function sha1 (data, base64 = false) {
  return crypto.createHash('sha1')
    .update(data, 'utf8')
    .digest(base64 ? 'base64' : 'hex')
}
