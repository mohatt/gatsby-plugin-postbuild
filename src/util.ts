import path from 'path'
import crypto from 'crypto'
import { mergeWith, isArray } from 'lodash'
import Debug from 'debug'
import { defaults, IPluginOptions } from './options'
import {
  GatsbyNodeHelpers,
  GatsbyPluginOptions,
  GatsbyReporterErrorMap,
  GatsbyReporterErrorMeta
} from './gatsby'

/**
 * Options interface
 */
export interface IOptions extends IPluginOptions {
  _plugin: string
  _root: string
  _public: string
  _pathPrefix: string
}

/**
 * Global options object
 */
export let options: IOptions = {
  _plugin: 'gatsby-plugin-postbuild',
  _root: '',
  _public: '',
  _pathPrefix: '',
  ...defaults
}

/**
 * Initializes plguin options and other utilities
 */
export function bootstrap ({ pluginOptions, gatsby }: {
  pluginOptions: GatsbyPluginOptions
  gatsby: GatsbyNodeHelpers
}): void {
  // Merge user-defined options with defaults
  options = mergeWith(options, pluginOptions,
    // Ensure arrays aren't merged by index
    (to, from) => isArray(from) ? from : undefined
  )

  // set private options
  options._root = gatsby.store.getState().program.directory
  options._public = path.join(options._root, 'public')
  options._pathPrefix = gatsby.pathPrefix

  // Initialize reporter
  const reporter = gatsby.reporter
  const errorMap: GatsbyReporterErrorMap = {
    10000: {
      text: (context: { message: string }) => {
        return `${options._plugin} threw an error while running:\n ${context.message}`
      },
      level: 'ERROR',
      type: 'PLUGIN'
    }
  }
  if (typeof reporter.setErrorMap === 'function') {
    reporter.setErrorMap(errorMap)
  }

  // Debug final options
  debug('Plugin initialized', options)
}

/**
 * Creates a structured error object to pass to gatsby's reporter
 */
export function createGatsbyError (message: string|Object, error?: Error, code = 10000): GatsbyReporterErrorMeta {
  return {
    id: `${options._plugin}_${code}`,
    context: typeof message === 'object'
      ? message
      : { message },
    error: error instanceof Error
      ? error
      : undefined
  }
}

/**
 * Prints a debug message
 */
export function debug (message: string, ...params: any[]): void {
  createDebug()(message, ...params)
}

/**
 * Creates a debug instance for the given namespace
 *
 * @param {string} namespace
 * @return {Function}
 */
export function createDebug (namespace: string = ''): Function {
  let ns = options._plugin
  if (namespace !== '') {
    ns += ':' + namespace
  }

  const debug = Debug(ns)
  return function (message: string, ...params: any[]) {
    message = params.length > 0
      ? `${message} %O \n\n`
      : `${message} \n\n`
    debug(message, ...params)
  }
}

/**
 * Creates a sha1 hash from a string.
 */
export function sha1 (data: string, base64 = false): string {
  return crypto.createHash('sha1')
    .update(data, 'utf8')
    .digest(base64 ? 'base64' : 'hex')
}

/**
 * Returns the file extension or checks if
 * a file has a given extension
 */
export function extName (file: string, checkExt?: string): string|boolean {
  const ext = path.extname(file).replace('.', '').toLowerCase()
  if (checkExt === undefined) return ext
  return ext === checkExt
}
