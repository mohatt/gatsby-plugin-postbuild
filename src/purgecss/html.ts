import { Promise } from 'bluebird'
import { promises as fs } from 'fs'
import path from 'path'
import parse5 from 'parse5'
import * as htmlparser2 from 'parse5-htmlparser2-tree-adapter'
import _ from 'lodash'
import { Purger, PurgeResult } from './purger'
import { FileWriter } from './writer'
import { AssetMapper } from './mapper'
import { options, createDebug, sha1, extName } from '../util'
const debug = createDebug('purgecss/html')

export type INode = htmlparser2.Node
export type IElement = htmlparser2.Element
export type IElementText = htmlparser2.TextNode
export type INodeWalker = (node: INode) => boolean

/**
 * Recursively walks through a htmlparser2 tree invoking a callback
 *  on every node on the tree. If the callback returned false
 *  the node will be removed
 *
 *  Note: This function may mutate the node tree
 *
 * @param {INode} node
 * @param {INodeWalker} cb
 * @return {boolean} - Return value is used for recursion
 */
function walk (node: INode, cb: INodeWalker): boolean {
  if ('children' in node) {
    if (node.children.length === 0) return cb(node)
    if (!cb(node)) return false
    node.children = node.children.filter(childNode => {
      return walk(childNode, cb)
    })
    return true
  }
  return cb(node)
}

/**
 * Interface for style objects
 */
export abstract class IStyle {
  /**
   * Style ID which should be set for shared styles
   */
  protected id?: string

  /**
   * Sets the id of the style object
   */
  setID (id: string): void {
    this.id = id
  }

  /**
   * Reads the id of the style object
   */
  getID (): string {
    return this.id !== undefined ? this.id : ''
  }

  /**
   * Check if the style object has an id
   */
  hasID (): boolean {
    return this.id !== undefined
  }

  /**
   * Reads the original css string
   */
  abstract read (): Promise<string>|string

  /**
   * Writes the optimized css string
   */
  abstract write (result: PurgeResult): Promise<void>|void
}

/**
 * Handles inline style nodes
 */
export class StyleInline extends IStyle {
  private textNode: IElementText
  constructor (node: IElementText) {
    super()
    this.textNode = node
  }

  /**
   * @inheritDoc
   */
  read (): string {
    return this.textNode.data
  }

  /**
   * @inheritDoc
   */
  write (result: PurgeResult): void {
    this.textNode.data = result.css
  }
}

/**
 * Handles external css files
 */
export class StyleFile extends IStyle {
  private readonly writer: FileWriter
  private readonly path: string
  constructor (writer: FileWriter, path: string) {
    super()
    this.writer = writer
    this.path = path
  }

  /**
   * @inheritDoc
   */
  async read (): Promise<string> {
    await fs.access(this.path)
    return await fs.readFile(this.path, 'utf-8')
  }

  /**
   * @inheritDoc
   */
  async write (result: PurgeResult): Promise<void> {
    await this.writer.write(this.path, result.css, result.rejected)
  }
}

/**
 * Handles html parsing/serialization and asset searching
 */
export class HtmlFile {
  /**
   * Path to the html file
   */
  path: string

  /**
   * Compiled ast tree
   */
  tree: htmlparser2.Document

  /**
   * Html string without CSS
   */
  nakedHtml: string

  /**
   * List of styles found in html with their metadata
   */
  styles: IStyle[] = []

  /**
   * Absolute paths to scripts found in html
   */
  scripts: string[] = []

  private readonly mapper
  private readonly purger
  private readonly writer

  /**
   * Creates a new instance for the given file path
   */
  constructor (path: string, mapper: AssetMapper, purger: Purger, writer: FileWriter) {
    this.path = path
    this.tree = htmlparser2.createDocument()
    this.nakedHtml = ''
    this.mapper = mapper
    this.purger = purger
    this.writer = writer
  }

  /**
   * Parses the html file into ast and writes
   * a style-less copy of it
   */
  async load (): Promise<void> {
    const filedata = await fs.readFile(this.path, 'utf8')
    this.tree = parse5.parse(filedata, {
      treeAdapter: htmlparser2
    })
    // clone the tree and remove all <style> nodes
    const nakedTree = _.cloneDeep(this.tree)
    walk(nakedTree, node => node.name !== 'style')
    // then serialize it back to HTML
    this.nakedHtml = parse5.serialize(nakedTree, {
      treeAdapter: htmlparser2
    })
    debug('Loaded html file', this.path)
  }

  /**
   * Searches for css/js assets in ast
   */
  loadAssets (): void {
    let inHead = true
    walk(this.tree, node => {
      if (node.name === 'body') {
        inHead = false
      } else if (node.name === 'style') {
        this.addStyle(node, inHead)
      } else if (node.name === 'link') {
        const rel = 'rel' in node.attribs && node.attribs.rel.toLowerCase().trim()
        if (rel === 'stylesheet') {
          this.addStyle(node, inHead)
        }
      } else if (node.name === 'script') {
        this.addScript(node)
      }
      return true
    })
    debug('Loaded html file assets', _.pick(this, ['path', 'styles', 'scripts']))
  }

  /**
   * Creates a style object for the given <style> or <link> node
   * then appends it to {self.styles}
   */
  addStyle (node: IElement, head = false): void {
    if (node.name === 'link') {
      const stylePath = this.resolveHref(node)
      if (typeof stylePath !== 'string' ||
        extName(stylePath, 'css') === false ||
        this.mapper.shouldIgnoreFile(stylePath)) {
        return
      }
      const style = new StyleFile(this.writer, stylePath)
      style.setID(stylePath)
      this.mapper.linkStyleToFile(stylePath, this)
      this.styles.push(style)
      return
    }

    if (node.children.length === 0) return
    const style = new StyleInline(node.children[0] as IElementText)
    if (head) {
      const id = 'id' in node.attribs && node.attribs.id !== ''
        ? node.attribs.id
        : sha1(style.read(), true)
      style.setID(id)
      this.mapper.linkStyleToFile(id, this)
    }
    this.styles.push(style)
  }

  /**
   * Creates a script filename from the given <script> node
   * checks if the script should be ignored, then appends
   * it to {self.scripts}
   */
  addScript (node: IElement): void {
    const scriptPath = this.resolveHref(node, 'src')
    if (typeof scriptPath !== 'string' ||
      extName(scriptPath, 'js') === false ||
      this.mapper.shouldIgnoreFile(scriptPath)) {
      return
    }
    this.scripts.push(scriptPath)
  }

  /**
   * Resolves a href attribute of a given node into local file path
   *
   * @param node - Element node object
   * @param attrib - name of the attribute
   * @return resolved file path or false on failure
   */
  resolveHref (node: IElement, attrib = 'href'): string|boolean {
    if (!(attrib in node.attribs)) return false
    let href = node.attribs[attrib].trim()
    if (href === '' || /^\w+:\/\//.test(href)) return false
    const prefix = options._pathPrefix
    if (prefix !== '') {
      if (href.indexOf(prefix) === 0) {
        href = href.replace(prefix, '')
      } else if (path.isAbsolute(href)) {
        return false
      }
    }

    if (path.isAbsolute(href)) {
      return path.join(options._public, href)
    }
    return path.resolve(path.dirname(this.path), href)
  }

  /**
   * Purges all style nodes then writes the final
   * optimized html file
   */
  purgeStyles (): Promise<void> {
    debug('Purging styles for html file', this.path)
    return Promise.mapSeries(this.styles, style => {
      return this.purger.purge(style, this)
    }).then(async result => {
      debug('Serializing purged html tree for', this.path)
      const ndata = parse5.serialize(this.tree, { treeAdapter: htmlparser2 })
      return await this.writer.write(
        this.path,
        ndata,
        options.purgecss.rejected ? result.flat() : null
      )
    })
  }
}
