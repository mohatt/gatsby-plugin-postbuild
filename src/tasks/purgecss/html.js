import { Promise } from 'bluebird'
import { promises as fs } from 'fs'
import path from 'path'
import parse5 from 'parse5'
import * as htmlparser2 from 'parse5-htmlparser2-tree-adapter'
import _ from 'lodash'
import { options, sha1 } from '../../util'

/**
 * Recursively walks through a htmlparser2 tree invoking a callback
 *  on every node on the tree. If the callback returned false
 *  the node will be removed
 *
 *  Note: This function may mutate the node tree
 *
 * @param {(htmlparser2.Node|htmlparser2.Element)} node
 * @param {function} cb
 * @return {boolean} - Return value is used for recursion
 */
function walk (node, cb) {
  if (!node.children) return cb(node)
  if (!cb(node)) return false
  node.children = node.children.filter(childNode => {
    return walk(childNode, cb)
  })
  return true
}

/**
 * Handles html parsing/serialization and asset searching
 */
export class HtmlFile {
  /**
   * Path to the html file
   * @type {string}
   */
  path

  /**
   * Compiled ast tree
   * @type {htmlparser2.Document}
   */
  tree

  /**
   * Html string without CSS
   * @type {string}
   */
  nakedHtml

  /**
   * List of styles found in html with their metadata
   * @type {[Object]}
   */
  styles = []

  /**
   * Absolute paths to scripts found in html
   * @type {[string]}
   */
  scripts = []

  /**
   * @type {Purger}
   */
  purger

  /**
   * @type {FileWriter}
   */
  writer

  /**
   * Creates a new instance for the given file path
   *
   * @constructor
   * @param {string} path
   * @param {Purger} purger
   * @param {FileWriter} writer
   */
  constructor (path, purger, writer) {
    this.path = path
    this.purger = purger
    this.writer = writer
  }

  /**
   * Parses the html file into ast and writes
   * a style-less copy of it
   *
   * @return {Promise<void>}
   */
  async load () {
    const filedata = await fs.readFile(this.path, 'utf8')
    this.tree = parse5.parse(filedata, {
      treeAdapter: htmlparser2
    })
    // clone the tree and remove all <style> nodes
    const nakedTree = _.cloneDeep(this.tree)
    walk(nakedTree, node => node.name !== 'style')
    // then serialize it back to HTML
    this.nakedHtml = parse5.serialize(nakedTree, {
      treeAdapter: htmlparser2
    })
  }

  /**
   * Searches for css/js assets in ast
   */
  loadAssets () {
    let inHead = true
    walk(this.tree, node => {
      if (node.name === 'body') {
        inHead = false
      } else if (node.name === 'style') {
        this.addStyle(node, inHead)
      } else if (node.name === 'link') {
        const rel = (node.attribs.rel || '').toLowerCase().trim()
        if (rel === 'stylesheet') {
          this.addStyle(node, inHead)
        }
      } else if (node.name === 'script') {
        this.addScript(node)
      }
      return true
    })
  }

  /**
   * Creates a style object for the given <style> or <link> node
   * then appends it to {self.styles}
   *
   * @param {htmlparser2.Element} node
   * @param {Boolean} head - whether this node was found under <head>
   */
  addStyle (node, head = false) {
    const style = {
      id: null,
      type: node.name,
      text: null,
      file: null
    }
    if (node.name === 'style') {
      if (node.children.length === 0) return
      style.text = node.children[0]
      style.text.data = style.text.data.trim()
      if (style.text.data === '') {
        return
      }
      if (head) {
        style.id = (node.attribs.id || '').trim() || sha1(style.text.data, true)
        this.purger.linkStyle(style, this)
      }
    } else if (node.name === 'link') {
      style.file = this.resolveHref(node)
      if (!style.file) {
        return
      }
      style.id = style.file
      this.purger.linkStyle(style, this)
    }
    this.styles.push(style)
  }

  /**
   * Creates a script filename from the given <script> node
   * checks if the script should be ignored, then appends
   * it to {self.scripts}
   *
   * @param {htmlparser2.Element} node
   */
  addScript (node) {
    const filename = this.resolveHref(node, 'src')
    if (!filename) {
      return
    }
    const relFilename = path.relative(options._public, filename)
    if (this.purger.shouldIgnoreScript(relFilename)) {
      return
    }
    this.scripts.push(filename)
  }

  /**
   * Resolves a href attribute of a given node into local file path
   *
   * @param {htmlparser2.Element} node
   * @param {string} attrib - name of the attribute
   * @return {string|boolean} - resolved file path or false on failure
   */
  resolveHref (node, attrib = 'href') {
    let href = (node.attribs[attrib] || '').trim()
    if (!href || /^\w+:\/\//.test(href)) return false
    const prefix = options._pathPrefix
    if (prefix) {
      if (href.indexOf(prefix) === 0) {
        href = href.replace(prefix, '')
      } else if (path.isAbsolute(href)) {
        return false
      }
    }

    if (path.isAbsolute(href)) {
      return path.join(options._public, href)
    }
    return path.resolve(path.dirname(this.path), href)
  }

  /**
   * Purges all style nodes then writes the final
   * optimized html file
   *
   * @return {Promise<void>}
   */
  purgeStyles () {
    return Promise.mapSeries(this.styles, style => {
      return this.purger.purge(style, this)
    }).then(result => {
      const ndata = parse5.serialize(this.tree, { treeAdapter: htmlparser2 })
      return this.writer.write(
        this.path,
        ndata,
        options.purgecss.rejected ? result.flat() : null
      )
    })
  }
}
