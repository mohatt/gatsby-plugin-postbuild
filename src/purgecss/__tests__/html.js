import { options } from '../../util'
import { mountFile, mountOptions } from '../../../test/utils'
import { HtmlFile } from '../html'

jest.mock('fs')
mountOptions()

describe('HtmlFile', () => {
  const { AssetMapper } = jest.createMockFromModule('../mapper')
  const { FileWriter } = jest.createMockFromModule('../writer')
  const { Purger } = jest.createMockFromModule('../purger')

  const mapper = new AssetMapper()
  const writer = new FileWriter()
  const purger = new Purger()

  it('correctly loads html file', async () => {
    let file
    const filename = '/path/to/file.html'
    const html = `
<html>
  <head>
    <style id="global" type="text/css">html{}</style>
    <style type="text/css">body{}</style>
    <link rel="stylesheet" href="styles/style.css" />
    <link rel="preload" href="scrips/foo.js" />
    <script src="scrips/bar.js"></script>
  </head>
  <body>
    <style type="text/css">div{}</style>
    <div>
      <style type="text/css">p{}</style>
      Hi
    </div>
    <script src="scrips/foo.js"></script>
  </body>
</html>`
    mountFile(filename, html)

    await expect(() => {
      file = new HtmlFile(filename, mapper, purger, writer)
      return file.load()
    }).not.toThrow()
    expect(file.nakedHtml).toMatchSnapshot()
    expect(() => { file.loadAssets() }).not.toThrow()
    expect(file.styles.length).toBe(5)
    expect(file.scripts.length).toBe(2)
    expect(mapper.linkStyleToFile.mock.calls.length).toBe(3)
    await expect(() => file.purgeStyles()).not.toThrow()
    expect(purger.purge.mock.calls.length).toBe(5)
    expect(writer.write.mock.calls.length).toBe(1)
  })

  it('correctly resolves asset hrefs', () => {
    let file
    const filename = `${options._public}/page/index.html`
    const hrefs = [
      'foo.js',
      './foo.js',
      '../foo.js',
      '/foo.js',
      '/dir/foo.js',
      '../dir/foo.js',
      '/dir/foo.js'
    ]
    const resolveHref = (href) => {
      return file.resolveHref({
        attribs: { href: href }
      })
    }

    expect(() => { file = new HtmlFile(filename) }).not.toThrow()
    expect(hrefs.map(h => resolveHref(h))).toMatchSnapshot()
    options._pathPrefix = '/path/prefix/'
    expect(hrefs.map(h => resolveHref(h))).toMatchSnapshot()
  })
})