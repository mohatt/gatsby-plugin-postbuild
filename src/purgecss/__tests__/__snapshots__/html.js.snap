// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtmlFile correctly loads html file 1`] = `
"<html><head>
    
    
    <link rel=\\"stylesheet\\" href=\\"styles/style.css\\">
    <link rel=\\"preload\\" href=\\"scrips/foo.js\\">
    <script src=\\"scrips/bar.js\\"></script>
  </head>
  <body>
    
    <div>
      
      Hi
    </div>
    <script src=\\"scrips/foo.js\\"></script>
  
</body></html>"
`;

exports[`HtmlFile correctly resolves asset hrefs 1`] = `
Array [
  "/virtual/project/public/page/foo.js",
  "/virtual/project/public/page/foo.js",
  "/virtual/project/public/foo.js",
  "/virtual/project/public/foo.js",
  "/virtual/project/public/dir/foo.js",
  "/virtual/project/public/dir/foo.js",
  "/virtual/project/public/dir/foo.js",
]
`;

exports[`HtmlFile correctly resolves asset hrefs 2`] = `
Array [
  "/virtual/project/public/page/foo.js",
  "/virtual/project/public/page/foo.js",
  "/virtual/project/public/foo.js",
  false,
  false,
  "/virtual/project/public/dir/foo.js",
  false,
]
`;
